include/xframe/xaxis_default.hpp

```diff
@@ -0,0 +1,326 @@
+/***************************************************************************
+* Copyright (c) 2017, Johan Mabille, Sylvain Corlay, Wolf Vollprecht and   *
+* Martin Renou                                                             *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XFRAME_XAXIS_DEFAULT_HPP
+#define XFRAME_XAXIS_DEFAULT_HPP
+
+#include <utility>
+#include <vector>
+#include <ostream>
+#include <iterator>
+
+#include "xtl/xiterator_base.hpp"
+
+#include "xaxis_base.hpp"
+#include "xaxis.hpp"
+
+namespace xf
+{
+    template <class L, class T>
+    class xaxis_default_iterator;
+
+    /*****************
+     * xaxis_default *
+     *****************/
+
+    template <class L, class T>
+    class xaxis_default : public xaxis_base<xaxis_default<L, T>>
+    {
+    public:
+
+        using base_type = xaxis_base<xaxis_default>;
+        using self_type = xaxis_default<L, T>;
+        using axis_type = xaxis<L, T>;
+        using key_type = typename base_type::key_type;
+        using label_list = std::vector<key_type>;
+        using mapped_type = typename base_type::mapped_type;
+        using size_type = typename base_type::size_type;
+        using difference_type = typename base_type::difference_type;
+        using iterator = typename base_type::iterator;
+        using iterator_value_type = std::pair<key_type, mapped_type>;
+        using const_iterator = typename base_type::const_iterator;
+        using reverse_iterator = typename base_type::reverse_iterator;
+        using const_reverse_iterator = typename base_type::const_reverse_iterator;
+
+        static_assert(std::is_integral<T>::value, "key_type L must be an integral type");
+
+        explicit xaxis_default(const size_type& size = 0);
+
+        bool is_sorted() const noexcept;
+
+        bool empty() const noexcept;
+        const size_type size() const noexcept;
```

Comment:
no need to return a temporary as `const`.

I think `empty` and `size` can be implemented in the `xaxis_base` class only since you'll store the labels, even for default axis.